apps/web/src/hooks/usePaymaster.ts

```diff
@@ -0,0 +1,126 @@
+import isZero from '@pancakeswap/utils/isZero'
+import { useActiveChainId } from 'hooks/useActiveChainId'
+import { useAtomValue } from 'jotai'
+import { useMemo } from 'react'
+import { Address, Hex, hexToBigInt, isAddress, stringify } from 'viem'
+import { serializeTransaction } from 'viem/zksync'
+
+import { ChainId } from '@pancakeswap/chains'
+import { useWalletClient } from 'wagmi'
+import { feeTokenAtom } from '../state/paymaster/atoms'
+import { getEip712Domain } from '../utils/paymaster'
+
+interface SwapCall {
+  address: Address
+  calldata: Hex
+  value: Hex
+}
+
+/**
+ * Zyfi Paymaster for the zkSync chain
+ */
+export const usePaymaster = () => {
+  const chain = useActiveChainId()
+  const { data: walletClient } = useWalletClient()
+
+  const feeToken = useAtomValue(feeTokenAtom)
+
+  /**
+   * Check if the Paymaster for zkSync is available
+   */
+  const isPaymasterAvailable = useMemo(() => {
+    return chain && chain.chainId === ChainId.ZKSYNC
+  }, [chain])
+
+  /**
+   * Check if a paymaster token is selected.
+   * Default is the native token to pay gas
+   */
+  const isPaymasterTokenActive = useMemo(() => {
+    return feeToken.isToken && feeToken.address && isAddress(feeToken.address)
+  }, [feeToken])
+
+  async function sendPaymasterTransaction(
+    call: SwapCall & {
+      gas?: string | bigint | undefined
+    },
+    account: Address,
+  ) {
+    if (!feeToken.isToken)
+      return Promise.reject(new Error('Selected gas token is not an ERC20 token. Unsupported by Paymaster.'))
+
+    const response = await fetch(`https://api.zyfi.org/api/erc20_paymaster/v1`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: stringify({
+        feeTokenAddress: feeToken.address,
+        gasLimit: call.gas,
+        txData: {
+          from: account,
+          to: call.address,
+          value: call.value,
+          data: call.calldata,
+        },
+      }),
+    })
+
+    if (response.ok) {
```

Comment:
personally i'd prefer to deal with whatever simpler first, so there will be fewer nested brackets
```typescript
if (!response.ok) {
  throw Error
}
// then proceed with the normal flow
```